### PR TITLE
[RHELC-899] Prevent packages marked for update to reinstall

### PR DESCRIPTION
**Pull request by r0x0d**
During our transaction handler, some packages were trying to be reinstalled/downgraded on the system without necessity, as they were already marked for update.

This patch prevents those packages marked for update to be reinstalled or downgraded in further cases.

Jira Issues: [RHELC-899](https://issues.redhat.com/browse/RHELC-899)

Checklist

- [x] PR has been tested manually in a VM (either author or reviewer)
- [x] Jira issue has been made public if possible
- [x] `[RHELC-]` is part of the PR title 
- [x] GitHub label has been added to help with Release notes 
- [x] PR title explains the change from the user's point of view
- [x] Code and tests are documented properly
- [x] The commits are squashed to as few commits as possible (without losing data) 
- [ ] When merged: Jira issue has been updated to `Release Pending` if relevant

### CODE REVIEW

**Comment by r0x0d** [comment]
This PR still needs some fixes for the DNF case.

Currently waiting for some response from the DNF folks to see what we can do in that front.

**Comment by abadger** [comment]
Looked at what's here now (the yum portion) and it looks good to me.

**Review by bocekm — APPROVED** [review]
@danmyway, from my POV it'd be good to have a regression-preventing integration test that would make sure that when following the RHELC-899 reproducer, convert2rhel installs the latest shim package available.

This PR covers well the https://issues.redhat.com/browse/RHELC-899 which is about yum. dnf has the same issue but it will be covered under a separate PR linked to [RHELC-920](https://issues.redhat.com/browse/RHELC-920).

File: convert2rhel/unit_tests/pkgmanager/handlers/dnf/dnf_test.py
Comment (r0x0d) [inline convert2rhel/unit_tests/pkgmanager/handlers/dnf/dnf_test.py:18]:
```suggestion
```

File: convert2rhel/pkgmanager/__init__.py
Comment (r0x0d) [inline convert2rhel/pkgmanager/__init__.py:44]:
```suggestion
```

File: convert2rhel/pkgmanager/handlers/dnf/__init__.py
Comment (r0x0d) [inline convert2rhel/pkgmanager/handlers/dnf/__init__.py:131]:
```suggestion
                self._base.upgrade(pkg_spec=pkg)
```

File: convert2rhel/pkgmanager/handlers/dnf/__init__.py
Comment (r0x0d) [inline convert2rhel/pkgmanager/handlers/dnf/__init__.py:135]:
```suggestion
                self._base.reinstall(pkg_spec=pkg)
```

File: convert2rhel/pkgmanager/handlers/dnf/__init__.py
Comment (r0x0d) [inline convert2rhel/pkgmanager/handlers/dnf/__init__.py:138]:
```suggestion
                    self._base.downgrade(pkg_spec=pkg)
```

File: convert2rhel/unit_tests/pkgmanager/handlers/dnf/dnf_test.py
Comment (r0x0d) [inline convert2rhel/unit_tests/pkgmanager/handlers/dnf/dnf_test.py]:
```suggestion
```

**Comment by r0x0d** [comment]
The failures here doesn't seem to be related to the code. Will trigger them again.

File: convert2rhel/pkgmanager/handlers/dnf/__init__.py
Comment (r0x0d) [reply convert2rhel/pkgmanager/handlers/dnf/__init__.py:21]: Not needed! Forgot to remove.

File: convert2rhel/pkgmanager/handlers/dnf/__init__.py
Comment (r0x0d) [reply convert2rhel/pkgmanager/handlers/dnf/__init__.py:123]: Updated locally to use rsplit instead, good catch!

**Review by abadger — APPROVED** [review]
Cool, @r0x0d has fixed the things I saw locally and they should show up in the next push.

File: tests/integration/tier1/single-yum-transaction/test_check_for_latest_packages.py
Comment (danmyway) [inline tests/integration/tier1/single-yum-transaction/test_check_for_latest_packages.py:10]:
```suggestion
from conftest import SYSTEM_RELEASE_ENV
from envparse import env
import subprocess

COMMAND = "curl -s 'https://yum.oracle.com/repo/OracleLinux/OL8/baseos/latest/x86_64/index.html' | grep -oP 'getPackage/\\Kshim-x64-[\\w.-]+(\\.x86_64)' | tail -1"

LATEST_IN_OL_REPO = subprocess.check_output(COMMAND, shell=True, text=True).strip()

PACKAGES_TO_VERIFY_RHEL_7 = [("shim-x64", "shim-x64-15.6-3.el7_9.x86_64")]
# Match the latest version used for conversion, currently Oracle Linux 8.7
PACKAGES_TO_VERIFY_RHEL_8_X = [("shim-x64", LATEST_IN_OL_REPO)]
# Mainly for CentOS Linux 8.5
PACKAGES_TO_VERIFY_RHEL_8_5 = [("shim-x64", "shim-x64-15.4-2.el8_1.x86_64")]

```

Comment (r0x0d) [reply tests/integration/tier1/single-yum-transaction/test_check_for_latest_packages.py:10]: good point on that. I will apply the change once we have results from tier1

File: tests/integration/tier1/single-yum-transaction/test_check_for_latest_packages.py
Comment (danmyway) [inline tests/integration/tier1/single-yum-transaction/test_check_for_latest_packages.py:12]: Please add the marker, also add to `pytest.ini`, please.
```suggestion

@pytest.mark.test_packages_upgraded_after_conversion
def test_packages_upgraded_after_conversion(convert2rhel, shell):
```

File: tests/integration/tier1/single-yum-transaction/test_check_for_latest_packages.py
Comment (danmyway) [reply tests/integration/tier1/single-yum-transaction/test_check_for_latest_packages.py:10]: Possibly `repoquery` would be better

File: tests/integration/tier1/single-yum-transaction/test_check_for_latest_packages.py
Comment (kokesak) [inline tests/integration/tier1/single-yum-transaction/test_check_for_latest_packages.py:53]:
```suggestion
        # If it is possible to update the package, the yum returncode yields 100
```

Comment (danmyway) [reply tests/integration/tier1/single-yum-transaction/test_check_for_latest_packages.py:53]: Will update in separate PR to unblock merge

**Comment by danmyway** [comment]
Just waiting for CentOS 8.5 tier1 to pass
Other tests are fine, results available here:
http://artifacts.osci.redhat.com/testing-farm/cf0d77ed-cccc-4903-8fc0-65625d48a545/
http://artifacts.osci.redhat.com/testing-farm/ca48e484-3bff-4fd2-ad1f-eb1ef468cf44/
http://artifacts.osci.redhat.com/testing-farm/ac4729e4-fafc-4f23-a16d-083613b19709/
http://artifacts.osci.redhat.com/testing-farm/77eb375a-05da-4a30-b60f-87a38e1b4861/
http://artifacts.osci.redhat.com/testing-farm/87b28049-e2ac-4aa2-82bf-0443b070a005/
http://artifacts.osci.redhat.com/testing-farm/0ea8ea7e-8bfc-41ee-809c-d6fc5c0673dc/
http://artifacts.osci.redhat.com/testing-farm/3c97c6b8-ff8b-4dae-9fdd-8584876637f4/
http://artifacts.osci.redhat.com/testing-farm/2d591be0-d44f-4764-8346-b78a99c2abc3/

**Comment by danmyway** [comment]
Test fail unrelated, resolving conflict and merging
http://artifacts.osci.redhat.com/testing-farm/6e32f0f2-5304-416b-8582-d63d3b14d52a/

**Pull request merged**
